prometheus-collectors/class-user-stats-collector.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+namespace Automattic\VIP\Prometheus;
+
+use Automattic\VIP\Utils\Context;
+
+use Prometheus\Gauge;
+use Prometheus\RegistryInterface;
+use Two_Factor_Core;
+
+class User_Stats_Collector implements CollectorInterface {
+	private Gauge $users_gauge;
+
+	private string $blog_id;
+	private const MAX_USERS_FOR_COUNT     = 1_000_000;
+	private const MAX_USERS_FOR_2FA_STATS = 5_000;
+	private const METRIC_OPTION           = 'vip-prom-users';
+	public const TFA_STATUS_ENABLED       = 'enabled';
+	public const TFA_STATUS_DISABLED      = 'disabled';
+	public const TFA_STATUS_DEFAULT       = 'unknown';
+
+	public function initialize( RegistryInterface $registry ): void {
+		$this->blog_id = Plugin::get_instance()->get_site_label();
+
+		$this->users_gauge = $registry->getOrRegisterGauge(
+			'user',
+			'count',
+			'Number of users by role and 2FA status',
+			[ 'site_id', 'role', 'tfa_status' ]
+		);
+	}
+
+	public function collect_metrics(): void {
+		$metrics = get_option( self::METRIC_OPTION, [] );
+
+		if ( empty( $metrics ) ) {
+			return;
+		}
+
+		foreach ( $metrics as $role => $counts ) {
+			foreach ( $counts as $status => $count ) {
+				$this->users_gauge->set( $count, [ $this->blog_id, $role, $status ] );
+			}
+		}
+	}
+
+	public function process_metrics(): void {
+		global $wpdb;
+
+		if ( ! Context::is_wp_cli() ) {
+			return;
+		}
+
+		// Since we bunch everything together under a single label this logic wouldn't make any sense.
+		if ( is_multisite() && wp_count_sites()['all'] > Plugin::MAX_NETWORK_SITES ) {
+			return;
+		}
+
+		// count_users() is expensive so skip collecting this metric if there are too many users.
+		// https://developer.wordpress.org/reference/functions/count_users/
+		if ( get_user_count() > self::MAX_USERS_FOR_COUNT ) {
```

Comment:
So the expense comes from the `avail_roles` calculation in the `count_users()` function? Is that why we're using the `get_user_count()` ?

Reply:
Yep, `get_user_count()` is a cached count of all users on the network so we can avoid collecting this metric altogether if it has an huge number of users.